File: Grind75/9-FloodFill/cpp/FloodFill.cpp
#include "FloodFill.h"
#include <queue>
#include <tuple>

std::vector<std::vector<int>> FloodFill::floodFill(
    std::vector<std::vector<int>>& image,
    int sr,
    int sc,
    int color) {

  int m = image.size();
  int n = (image.empty()) ? 0 : image[0].size();

  if (sr < 0 || sr >= m)
    return image;
  if (sc < 0 || sc >= n)
    return image;

  int origColor = image[sr][sc];
  std::queue<std::pair<int, int>> q;
  q.push(std::make_pair(sr, sc));

  while (!q.empty()) {
    std::pair<int, int> coord = q.front();
    int row = coord.first;
    int col = coord.second;
    image[row][col] = color;
    q.pop();
    if (row-1 >= 0 && image[row-1][col] == origColor)
      q.push(std::make_pair(row-1, col));
    if (row+1 < m && image[row+1][col] == origColor)
      q.push(std::make_pair(row+1, col));
    if (col-1 >= 0 && image[row][col-1] == origColor)
      q.push(std::make_pair(row, col-1));
    if (col+1 < n && image[row][col+1] == origColor)
      q.push(std::make_pair(row, col+1));
  }
  return image;
}

File: Grind75/9-FloodFill/python/FloodFill.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python/')
from List import *
from queue import Queue

class Solution:
  def floodFill(self, image: list[list[int]], sr: int, sc: int,
      color: int) -> list[list[int]]:
    if image is None or len(image) == 0 or len(image[0]) == 0:
      return image
    N = len(image)
    M = len(image[0])
    visited = [[False for m in range(M)] for n in range(N)]
    origColor = image[sr][sc]
    q = Queue()
    q.put((sr,sc))
    while not q.empty():
      (r,c) = q.get()
      visited[r][c] = True
      image[r][c] = color
      if (r-1 >= 0) and (image[r-1][c] == origColor) and (not visited[r-1][c]):
        q.put((r-1,c))
      if (r+1 < N)  and (image[r+1][c] == origColor) and (not visited[r+1][c]):
        q.put((r+1,c))
      if (c-1 >= 0) and (image[r][c-1] == origColor) and (not visited[r][c-1]):
        q.put((r,c-1))
      if (c+1 < M)  and (image[r][c+1] == origColor) and (not visited[r][c+1]):
        q.put((r,c+1))
    return image

if __name__ == '__main__':
  soln = Solution()

  testCases = [
      ([[1,1,1],[1,1,0],[1,0,1]], [[2,2,2],[2,2,0],[2,0,1]]),
      ]

  for test in testCases:
    for i in range(len(test[0])):
      print(test[0][i], end='   ')
      print(test[1][i])
    result = soln.floodFill(test[0], 1, 1, 2)
    for i in range(len(result)):
      print(result[i])

  print('TESTS PASSED!')

